### PR TITLE
fix(Module): allow passing template/templateUrl in array notation

### PR DESCRIPTION
**Pull request by shahata**

### CODE REVIEW

**Comment by gkalpak** [comment]
LGTM

File: src/loader.js [hunk 1]
```diff
@@ -362,7 +362,7 @@ function setupModuleLoader(window) {
           component: function(name, options) {
             function factory($injector) {
               function makeInjectable(fn) {
-                if (angular.isFunction(fn)) {
+                if (isFunction(fn) || isArray(fn)) {
```
Comment (petebacondarwin) [inline src/loader.js]: I am pretty sure we cannot rely on being in the closure of angular here.
Although the unit tests will pass, I don't think that `isFunction` and `isArray` will be available in practice.

Comment (shahata) [inline src/loader.js]: `isFunction` is used further down in the same file, so it means we're fine, no? https://github.com/angular/angular.js/blob/master/src/loader.js#L454

Comment (petebacondarwin) [inline src/loader.js]: I would like to see that working in practice before I agree... I wonder if that other use ever worked... :-)

Comment (petebacondarwin) [inline src/loader.js]: That usage was added by an external committer... https://github.com/angular/angular.js/commit/351fe4b79c50a45a11af2fcd2aa7b6fd3b70058d

Comment (shahata) [inline src/loader.js]: If that line didn't work every single service/directive/etc. you define would throw... (btw, it also has a bug of not handling array notation :smile:)

Comment (shahata) [inline src/loader.js]: I also checked in debugger to make sure that it is indeed in the closure

Comment (gkalpak) [inline src/loader.js]: It is also accessing `minErr` near the top of the file. It definitely relies on having access to some stuff. (Whether it _should_ is another story...)

Comment (shahata) [inline src/loader.js]: Ah, but maybe angular-loader.js will have a problem, I'll make sure

**Comment by Narretz** [comment]
I feel like this should be documented, no?

**Comment by Narretz** [comment]
Also, what's the reason why component() itself is not injectable?

**Comment by gkalpak** [comment]
`component()` does not accept a factory function, it gets the _CDO_ (Component Definition Object :smile:) directly.
Although, I've been thinking it might be convenient to overload it, in order to also accept a factory function.

**Comment by petebacondarwin** [comment]
I think it is nice that we are not injecting into the CDO. Anything that needs injection can ask for it.

File: src/loader.js (hunk 1 above)
Comment (lgalfaso) [inline src/loader.js]: `isFunction` will work because of https://github.com/angular/angular.js/blob/master/src/loader.prefix#L8 `isArray` will not work

Comment (shahata) [inline src/loader.js]: Does anybody have a strong opinion on whether to add `isArray` up there or just use `angular.isFunction` & `angular.isArray`? I don't really care

Comment (lgalfaso) [inline src/loader.js]: if you just do `angular.isArray` you would break the loader as it is meant to be run without angular

Comment (Narretz) [inline src/loader.js]: I think there's an open issue about the loader being broken for certain modules. Maybe that already happens?

Comment (petebacondarwin) [inline src/loader.js]: `Array.isArray` is supported on all the browsers that we support in (outside of v1.2) (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/isArray#Browser_compatibility) so should just use that in this case...

Comment (petebacondarwin) [inline src/loader.js]: We should definitely have a comment at the top of this file, explaining where we get `isFunction` from.

Comment (lgalfaso) [inline src/loader.js]: @petebacondarwin in 99% of the cases (that is when people use angular and not just the loader), `isFunction` comes from `src/Angular.js` and I doubt that there is a need to explain this. For the 1% of cases that the loader is used stand-alone, then it comes from another place.

**Comment by Narretz** [comment]
I'd still like to see a mention in the docs that you can you can inject any injectables into the template / templateUrl fn. This is far from obvious.

**Comment by lgalfaso** [comment]
@Narretz do you want to open an issue to track that?

**Comment by Narretz** [comment]
@lgalfaso I've added this to the docs just now, but I would really prefer if docs landed as part of the code changes.

**Pull request closed without merging**